### PR TITLE
feat(core): email relay

### PR DESCRIPTION
**Pull request by szkl**
### Description

TBD

### Related Issues

- Related #2415

### Testing

- [x] Manual

### Checklist

- [x] I have read the [CONTRIBUTING](CONTRIBUTING.md) guidelines
- [x] I have tested my code (manually and/or automated if applicable)
- [x] I have updated the documentation (if necessary)

### CODE REVIEW

**Comment by szkl** [comment]
Initial pull request was https://github.com/proofzero/rollupid/pull/2735

**Review by Cosmin-Parvulescu — COMMENTED** [review]
I'm not sure how to test this, so maybe we can schedule a pairing together tomorrow morning. Left some questions for then.

File: platform/email/src/types.ts [hunk 1]
```diff
@@ -10,14 +10,14 @@ export interface Environment {
 }
 
 /** CF EmailMessage type; not provided in CF types lib */
-export interface CloudflareEmailMessage<Body = unknown> {
+export interface CloudflareEmailMessage {
```
[the remark this answers is not in the document]
Comment (szkl) [reply platform/email/src/types.ts]: Not really but independent for now and I've removed this from the change set.

File: platform/core/src/relay.ts [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+  const email = await postalMime.parse(message)
+
+  const dkim: DKIM = {
+    dkim_domain: env.INTERNAL_RELAY_DKIM_DOMAIN,
+    dkim_selector: env.INTERNAL_RELAY_DKIM_SELECTOR,
+    dkim_private_key: env.SECRET_RELAY_DKIM_PRIVATE_KEY,
+  }
+
+  const recipients = new Array<Address>()
+    .concat(email.to || [])
```
[the remark this answers is not in the document]
Comment (szkl) [reply platform/core/src/relay.ts:43]: I guess it is better to be ready for any possible case.

File: platform/core/src/relay.ts [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+        ],
+      },
+      null,
+      2
+    ),
+  })
+
+  const response = await fetch(request)
+  if (!response.ok)
+    console.error(JSON.stringify(await response.json(), null, 2))
```
[the remark this answers is not in the document]
Comment (szkl) [reply platform/core/src/relay.ts:120]: Yes. MailChannels responds with JSON for errors but it is also possible to have any kind of failure. It should be guarded.

**Comment by Cosmin-Parvulescu** [comment]
- Tried attaching an image on Gmail web client, relayed e-mail lost attachment
- CC / BCC seems to work OK
- Tried attaching an XLS on Gmail web client, relayed e-mail lost attachment on Gmail, preserved on @kubelt.com

File: platform/core/.dev.vars.example [hunk 4]
```diff
@@ -3,6 +3,10 @@ SECRET_JWKS = '[]'
 
 INTERNAL_DKIM_SELECTOR = ""
 
+INTERNAL_RELAY_DKIM_DOMAIN = ""
+INTERNAL_RELAY_DKIM_SELECTOR = ""
```
[the remark this answers is not in the document]
Comment (szkl) [reply platform/core/.dev.vars.example:7]: No. I just put them here as an example for local development if needed.

File: .github/workflows/main-core.yaml [hunk 5]
```diff
@@ -62,6 +62,10 @@ jobs:
 
             INTERNAL_DKIM_SELECTOR
 
+            INTERNAL_RELAY_DKIM_DOMAIN
+            INTERNAL_RELAY_DKIM_SELECTOR
```
[the remark this answers is not in the document]
Comment (szkl) [reply .github/workflows/main-core.yaml:66]: Probably, not. I am going to check it.

File: platform/core/src/index.ts [hunk 6]
```diff
@@ -36,6 +37,17 @@ export default {
       },
     })
   },
+  async email(message: CloudflareEmailMessage, env: Environment) {
```
[the remark this answers is not in the document]
Comment (szkl) [reply platform/core/src/index.ts:40]: This method is called by the worker when an email is received.

File: platform/core/src/index.ts [hunk 7]
```diff
@@ -36,6 +37,17 @@ export default {
...
+    const decoder = new TextDecoder()
+    const reader = message.raw.getReader()
+    let content = ''
+    while (true) {
+      const { done, value } = await reader.read()
```
[the remark this answers is not in the document]
Comment (szkl) [reply platform/core/src/index.ts:45]: Good call. I'll lookup what kind of errors can be thrown by this call.

File: platform/core/src/relay.ts [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+  }
+
+  const recipients = new Array<Address>()
+    .concat(email.to || [])
+    .concat(email.cc || [])
+    .filter((recipient) =>
+      recipient.address.endsWith(`@${env.INTERNAL_RELAY_DKIM_DOMAIN}`)
+    )
+
+  for (const recipient of recipients) {
```
[the remark this answers is not in the document]
Comment (szkl) [reply platform/core/src/relay.ts:49]: Not really, not right away at least. This is only a background process.

File: platform/core/src/index.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (szkl) [reply platform/core/src/index.ts:45]: I haven't found anything useful.

**Review by Cosmin-Parvulescu — APPROVED** [review]

**Pull request merged**